### DISCUSSION TITLE
How to make import?

### DISCUSSION THREAD

**Discussion by Tasyk1903** [filamentphp/filament#3225] [Help]
Hello, I've been scratching my head over importing for several days, please help. How to upload a file and pass it to the controller?

`class ListProducts extends ListRecords implements HasForms
{
    use InteractsWithForms;

    protected static string $resource = ProductResource::class;

    protected function getActions(): array
    {
        return [
            Actions\CreateAction::make(),
        ];
    }

    protected function getTableHeaderActions(): array
    {
        return array_merge(parent::getTableHeaderActions(), [
            Action::make('Import')
                ->action(function (Request $request) {
                })
                ->form([
                    Forms\Components\FileUpload::make('attachment')
                        ->required()
                ])
    ]);
    }
}`

My controller
`class ExportImportController extends Controller
{
  
    public static function import(Request $request)
    {

        Excel::import(new \App\Imports\Products(), $request->file('import'));

        return redirect()->back()->with('success', 'Import Has Been updated successfully');
    }
}`

**Comment by danharrin** [comment]
```
$table
    ->headerActions([
        Action::make('Import')
            ->action(function (array $data) {
                Excel::import(new \App\Imports\Products(), $data['attachment']);

                Filament::notify('success', 'Import Has Been updated successfully');
            })
            ->form([
                Forms\Components\FileUpload::make('attachment')
                    ->required()
            ])
    ])
```

**Reply by danharrin** [reply]
or you can continue to override `getTableHeaderActions()`, but this is cleaner for resources.

**Reply by Tasyk1903** [reply]
Understood thanks. I was close to a solution
Now I'm getting this error, is it something to do with the name or path of the file?
Could not find zip member zip://D:\projects\storage\framework\cache\laravel-excel\laravel-excel-BCYKYhhu24PBxbxwECeuYyHigx6cKgOh.xlsx#_rels/.rels

**Reply by danharrin** [reply]
sounds like a problem with your import package, not Filament?

**Reply by Tasyk1903** [reply]
So it is, but if it was loaded through the controller itself, then everything worked

**Reply by Tasyk1903** [reply]
> sounds like a problem with your import package, not Filament?

Any ideas how to fix this?

**Reply by danharrin** [reply]
Check what path Filament returns, check what path your import package needs, do debugging. I cant really help you

**Reply by Tasyk1903** [reply]
> Check what path Filament returns, check what path your import package needs, do debugging. I cant really help you

filament returns null path and I can't get it

**Reply by danharrin** [reply]
Check the entire `$data` array?

**Reply by Tasyk1903** [reply]
> Check the entire `$data` array?

If through the controller, then it holds the file first in temporary storage. A filament immediately puts it in storage/app/public

**Reply by danharrin** [reply]
So you don't want the file saved to storage, you want the file object?

Try `->beforeStateDehydrated(null)` on the file upload field

**Reply by Tasyk1903** [reply]
> So you don't want the file saved to storage, you want the file object?
> 
> Try `->beforeStateDehydrated(null)` on the file upload field

In filament i get an empty array
![1](https://user-images.githubusercontent.com/109950187/180985872-b559131b-7c6c-4936-8874-070061003557.JPG)
If I do dd($data['import'], then I get just a link  {"C:\OpenServer\userdata\temp\php29EB.tmp"}

dd($request)  through controller
![2](https://user-images.githubusercontent.com/109950187/180986857-5709fada-0cc1-4706-bf1a-1e1ee3b71533.JPG)

**Reply by danharrin** [reply]
Are you trying to use the request to get Filament files? that wont work, you need to use the $data array

**Reply by danharrin** [reply]
if you use `beforeStateDehydrated(null)` on the file upload, you will get a TemporaryFileUpload

**Reply by Tasyk1903** [reply]
I'm finally with these) Thank you. The only point is, if the file has an error, how can you display these errors?
![Снимок](https://user-images.githubusercontent.com/109950187/181007713-d4064413-fd1a-4a52-88d4-a9ee6c652f95.JPG)

**Reply by pxlrbt** [reply]
Catch the error and send a different notification in that case.
